In [1]:
import pandas as pd 
import numpy as np 


In [6]:
df = pd.read_csv("mercedese_train.csv" , usecols=['X1', 'X2', 'X3', 'X4', 'X5', 'X6'])
df.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [7]:
# lets have look at how many labels are present in each category
for col in df.columns:
    print(col,': has',len(df[col].unique()),'columns')

X1 : has 27 columns
X2 : has 44 columns
X3 : has 7 columns
X4 : has 4 columns
X5 : has 29 columns
X6 : has 12 columns


In [8]:
df_OHE = pd.get_dummies(df,drop_first=True)
df_OHE

X1_aa  X1_ab  X1_b  X1_c  X1_d  X1_e  X1_f  X1_g  X1_h  X1_i  ...  X6_c  \
0         0      0     0     0     0     0     0     0     0     0  ...     0   
1         0      0     0     0     0     0     0     0     0     0  ...     0   
2         0      0     0     0     0     0     0     0     0     0  ...     0   
3         0      0     0     0     0     0     0     0     0     0  ...     0   
4         0      0     0     0     0     0     0     0     0     0  ...     0   
...     ...    ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
4204      0      0     0     0     0     0     0     0     0     0  ...     0   
4205      0      0     0     0     0     0     0     0     0     0  ...     0   
4206      0      0     0     0     0     0     0     0     0     0  ...     0   
4207      0      0     0     0     0     0     0     0     0     0  ...     0   
4208      0      0     0     0     0     0     0     0     0     0  ...     0   

      X6_d  X6_e  X6_f  X6_g  X6_h  X6_i  X6_j  X6_k  X6_l  
0        0     0     0     0     0     0     1     0     0  
1        0     0     0     0     0     0     0     0     1  
2        0     0     0     0     0     0     1     0     0  
3        0     0     0     0     0     0     0     0     1  
4        1     0     0     0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
4204     1     0     0     0     0     0     0     0     0  
4205     0     0     0     0     1     0     0     0     0  
4206     0     0     0     1     0     0     0     0     0  
4207     0     0     0     0     0     0     0     0     1  
4208     0     0     0     1     0     0     0     0     0  

[4209 rows x 117 columns]

### As here , we took only 5 columns and applied OHE , then it resulted into 117 columns . So imaginge if in real world we have more than 500 categories inside a single column.Then it wont be feasible . So to overcome this we are using approch of most frequent categories.


In [13]:
# lets find the most frequent labels in certain columns (let column = X2)
df['X2'].value_counts().sort_values(ascending = False).head(10)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
Name: X2, dtype: int64

### So above are the most frequent categories in column X2

In [17]:
top_10_labels = [x for x in df['X2'].value_counts().sort_values(ascending = False).head(10).index]
top_10_labels

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [21]:
for label in top_10_labels:
    df[label] = np.where(df['X2'] == label, 1, 0)

df[['X2']+top_10_labels].head(10)

X2  as  ae  ai  m  ak  r  n  s  f  e
0  at   0   0   0  0   0  0  0  0  0  0
1  av   0   0   0  0   0  0  0  0  0  0
2   n   0   0   0  0   0  0  1  0  0  0
3   n   0   0   0  0   0  0  1  0  0  0
4   n   0   0   0  0   0  0  1  0  0  0
5   e   0   0   0  0   0  0  0  0  0  1
6   e   0   0   0  0   0  0  0  0  0  1
7  as   1   0   0  0   0  0  0  0  0  0
8  as   1   0   0  0   0  0  0  0  0  0
9  aq   0   0   0  0   0  0  0  0  0  0

In [22]:
def one_hot_top(df , variable , top_10_labels):
    # function to create the dummy variables for most frequent labels inside the column
    # we can vary the number of most frequent labels
    
    for label in top_10_labels:
        df[variable+'_'+label] = np.where(df[variable] == label ,1 , 0 )
        
# read the data again
df = pd.read_csv("mercedese_train.csv",usecols=['X1','X2','X3','X4','X5','X6'])
one_hot_top(df,'X2',top_10_labels)
df.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0  v  at  a  d  u  j      0      0      0     0      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0      0     0     1     0   
3  t   n  f  d  x  l      0      0      0     0      0     0     1     0   
4  v   n  f  d  h  d      0      0      0     0      0     0     1     0   

   X2_f  X2_e  
0     0     0  
1     0     0  
2     0     0  
3     0     0  
4     0     0

In [23]:
top_10 = [x for x in df['X1'].value_counts().sort_values(ascending = False).head(10).index]

one_hot_top(df,'X1',top_10)

In [24]:
df.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X1_aa  X1_s  X1_b  X1_l  \
0  v  at  a  d  u  j      0      0      0     0  ...      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...      0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...      0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...      0     0     0     0   

   X1_v  X1_r  X1_i  X1_a  X1_c  X1_o  
0     1     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     0     0     0     0  
3     0     0     0     0     0     0  
4     1     0     0     0     0     0  

[5 rows x 26 columns]

In [25]:
top_10 = [x for x in df['X3'].value_counts().sort_values(ascending = False).head(10).index]

one_hot_top(df,'X3',top_10)
df.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X1_a  X1_c  X1_o  X3_c  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     1   
3  t   n  f  d  x  l      0      0      0     0  ...     0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     0     0     0     0   

   X3_f  X3_a  X3_d  X3_g  X3_e  X3_b  
0     0     1     0     0     0     0  
1     0     0     0     0     1     0  
2     0     0     0     0     0     0  
3     1     0     0     0     0     0  
4     1     0     0     0     0     0  

[5 rows x 33 columns]

In [26]:
top_10 = [x for x in df['X4'].value_counts().sort_values(ascending = False).head(10).index]

one_hot_top(df,'X4',top_10)
top_10 = [x for x in df['X5'].value_counts().sort_values(ascending = False).head(10).index]

one_hot_top(df,'X5',top_10)


In [27]:
df.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X5_v  X5_w  X5_q  X5_r  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...     0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     0     0     0     0   

   X5_s  X5_d  X5_n  X5_p  X5_m  X5_i  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     0     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 47 columns]

## Here the OHE columns are now 47 , way less than generated by keeping all the categories from each columns (117)

### Advantages :
- Straight forward to implement
- Does not expand massively the feature space

### Disadvantages:
- Does not add any information that adds variable value more
- Does not keep the information regarding ignored labeled data